File: Lua/problem_3/algorithm.lua
function maxPrimeFactors(n)
  local maxPrime = -1

  while n % 2 == 0 do
    maxPrime = 2
    n = n / 2
  end

  local i = 3
  while i <= math.sqrt(n) do
    while n % i == 0 do
      maxPrime = i
      n = n / i
    end
    i = i + 2
  end

  if n > 2 then
    maxPrime = n
  end

  return math.floor(maxPrime)
end

function main()
  local n = maxPrimeFactors(600851475143)
  print(n)
end

main()

File: C/problem_3/algorithm.c
#include <stdio.h>
#include <math.h>

long long maxPrimeFactors(long long n)
{
    long long maxPrime = -1;

    while (n % 2 == 0)
    {
        maxPrime = 2;
        n /= 2;
    }

    for (int i = 3; i <= sqrt(n); i += 2)
    {
        while (n % i == 0)
        {
            maxPrime = i;
            n /= i;
        }
    }

    if (n > 2)
    {
        maxPrime = n;
    }

    return maxPrime;
}

int main()
{
    long long n = maxPrimeFactors(600851475143);

    printf("%i\n", n);

    return 0;
}